src/core/connection.cpp

```diff
@@ -33,360 +33,631 @@
-
-std::unique_ptr<Connection>
-Connection::make(std::unique_ptr<websockets::WebSocketClient> p_websocket, std::unique_ptr<ErrorHandler> p_error_handler, WSFactory* wsFactory)
-{
-    assert(p_websocket);
-    assert(p_error_handler);
-
-    const unsigned MAX_NUM_REDIRECTIONS = 3;
-
-    for (unsigned num_redirections = 0; num_redirections < MAX_NUM_REDIRECTIONS;
-         ++num_redirections) {
-        p_websocket->set_receive_timeout(std::chrono::seconds(10));
-
-        const std::string uri = p_websocket->uri();
-
-        std::unique_ptr<Connection> p(new Connection(std::move(p_websocket), std::move(p_error_handler)));
-
-        Buffer buffer;
-        parser::MessageList messages;
-        State& state = p->state_;
-
-        if (p->receive_(&buffer, &messages) != websockets::State::OPEN)
-            return p;
-
-        if (messages.size() != 1 || state != State::CHALLENGING) {
-            p->close();
-            return p;
-        }
-
-        {
-            MessageBuilder chr(Topic::CONNECTION, Action::CHALLENGE_RESPONSE);
-            chr.add_argument(uri);
-
-            if (p->send_(chr) != websockets::State::OPEN)
-                return p;
-        }
-
-        if (p->receive_(&buffer, &messages) != websockets::State::OPEN)
-            return p;
+#ifndef NDEBUG
+#include <iostream>
+#define DEBUG_MSG(str) do { std::cout << str << std::endl; } while( false )
+#else
+#define DEBUG_MSG(str) do { } while ( false )
+#endif
 
-        if (messages.size() != 1) {
-            p->close();
-            return p;
-        }
-
-        if (state == State::AWAIT_AUTHENTICATION)
-            return p;
-
-        if (state == State::AWAIT_CONNECTION) {
-            const Message& redirect_msg = messages.front();
-            assert(redirect_msg.num_arguments() == 1);
-
-            const Buffer& uri_buffer = redirect_msg[0];
-            std::string new_uri(uri_buffer.cbegin(), uri_buffer.cend());
+namespace deepstream {
 
-            p_websocket = p->p_websocket_->construct(new_uri);
-            p_error_handler = std::move(p->p_error_handler_);
+    Connection::Connection(const std::string &uri, WSHandler &ws_handler,
+            ErrorHandler &error_handler, Event &event, Presence &presence)
+        : state_(ConnectionState::CLOSED)
+        , error_handler_(error_handler)
+        , ws_handler_(ws_handler)
+        , p_login_callback_(nullptr)
+        , p_auth_params_(nullptr)
+        , event_(event)
+        , presence_(presence)
+    {
+        assert(ws_handler.state() == WSState::CLOSED);
+
+        ws_handler.URI(uri);
+
+        using namespace std::placeholders;
+        ws_handler.on_message(std::bind(&Connection::on_message, this, _1));
+        ws_handler.on_error(std::bind(&Connection::on_error, this, _1));
+        ws_handler.on_open(std::bind(&Connection::on_open, this));
+        ws_handler.on_close(std::bind(&Connection::on_close, this));
+
+        ws_handler.open();
+    }
 
-            continue;
+    /*
```

Comment:
absolutely. i left the comments in to make it easier for me to check whether I'd missed any features that were in the client before but forgot to remove them